### ISSUE TITLE
[ioctl_map] Page mapping kernel request failed (ptr=0x7fcef2820000, n_pages=1): Invalid argument, when running nvm-block-bench

### ISSUE THREAD

**Issue by AkhilSrinivasSolidigm** [ZaidQureshi/bam#42]
**Describe the bug**

[ioctl_map] Page mapping kernel request failed (ptr=0x7fcef2820000, n_pages=1): Invalid argument, when running nvm-block-bench
when running 
sudo ./bin/nvm-block-bench --threads=262144 --blk_size=64 --reqs=1 --pages=262144 --queue_depth=1024  --page_size=512 --num_blks=2097152 --gpu=0 --n_ctrls=1 --num_queues=128 --random=false

**To Reproduce**
After successfully following all steps to install BaM with no errors, loading drive with make load

root@xxxx:~/GIDS/bam/build/module# ls -l /dev | grep libnvm
crwxrwxrwx   1 root root    506,   0 Oct 22 10:21 libnvm0

Drive loaded successfully
Command: dmesg | tail -10
Output: [  456.310326] Adding controller device: c8:00.0
[  456.310550] Character device /dev/libnvm0 created (506.0)
[  456.310610] libnvm helper loaded

Now running sudo ./bin/nvm-block-bench --threads=262144 --blk_size=64 --reqs=1 --pages=262144 --queue_depth=1024  --page_size=512 --num_blks=2097152 --gpu=0 --n_ctrls=1 --num_queues=128 --random=false

Gives [ioctl_map] Page mapping kernel request failed (ptr=0x7fcef2820000, n_pages=1): Invalid argument
Unexpected error: Failed to map device memory: Invalid argument

dmesg after failure:
Oct 22 10:31:15 kernel: [ 1028.550861] Unknown ioctl command from process 2389: 1075347458
Oct 22 10:31:54 kernel: [ 1067.493923] Unknown ioctl command from process 2400: 1075347458

**Expected behavior**
To run successfully without failures.

**Machine Setup (please complete the following information):**
 - OS : Ubuntu 20.04
 - NVIDIA Driver : 560.35.03, CUDA Versions: 12.4, GPU : L4
 - SSD used : Gen5 Solidigm PS1010
 - System: Supermicro H13SSw
 - RAM : 16 GB
 - CPU: AMD EPYC 9124 16-Core Processor

**Additional context**

Ensured nvcc and cuda are same version : 
root@xxxx:~/GIDS/bam/build# nvcc --version
nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2024 NVIDIA Corporation
Built on Thu_Mar_28_02:18:24_PDT_2024
Cuda compilation tools, release 12.4, V12.4.131
Build cuda_12.4.r12.4/compiler.34097967_0

uname -a :

Linux 5.8.0-050800-generic #202008022230 SMP Sun Aug 2 22:33:21 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux

**Comment by msharmavikram** [comment]
Is this a workstation or desktop system? If yes, this may be bit more tricky. 

I wonder if GPUdirect is supported in the platform. If it's not, BaM won't work.

**Comment by AkhilSrinivasSolidigm** [comment]
This is a workstation and have previously run GPUDirect but on a different OS. Is GDS and MLNX with NVMe a requirement for BaM support?

**Comment by msharmavikram** [comment]
No. GDS will work without GPUDirect in compact mode. BaM has no notion of compact mode.

**Comment by AkhilSrinivasSolidigm** [comment]
Is there any other logs or setup information I can provide to debug this issue? I've tried RHEL 9.2 as well with the same issue. Different drives too. Do you have any recommendations or feedback regarding the setup?

**Comment by msharmavikram** [comment]
Unsure as this is entering a specific system-specific feature and something that we never accounted for. 

We only know what it takes based on what we described here - https://github.com/ZaidQureshi/bam?tab=readme-ov-file#hardwaresystem-requirements 

We cannot support all the universal system configurations. Your team member Wayne already has brought up BaM and would recommend working closely with him.

**Comment by AkhilSrinivasSolidigm** [comment]
I've had a word with Wayne and looks like he hasn't encountered this issue to help us with the debug. There is similar closed issue https://github.com/ZaidQureshi/bam/issues/26 with no resolution other than RTX series not supported. I believe we have met the minimum and recommended requirements for BaM and seeing this issue. Do you have any suggestions on our hardware setup, are we missing anything that could likely be causing this issue?

**Comment by AkhilSrinivasSolidigm** [comment]
Bumping again to get some traction on this issue. Any suggestions on this issue would be greatly helpful in our setup.

**Comment by starry12** [comment]
Hi, I encountered the same issue when configuring on the new environment and successfully resolved it.
I think the main reason is the incompatibility between the linux kernel version and the nvidia driver version.

There are two possible problems you should check for:
1. You need to recompile the Nvidia Driver Kernel Symbols for the current environment. This is also mentioned in the _readme_. The purpose is to ensure that the driver is compatible with the current kernel version of the system.
2. Check your /build/CMakeCache.txt file to make sure you're using the correct version of the nvidia driver. 
Like this:
```
//Path to a file
driver include:PATH=/usr/src/nvidia-550.90.12/nvidia
```
This needs to be the version you recompiled in 1.
If not, use this:
cmake -D NVIDIA=/usr/src/nvidia-550.54.14/ -D driver_include:PATH=/usr/src/nvidia-550.54.14/nvidia ..

Hope this helps you!

**Comment by msharmavikram** [comment]
Thanks for the comment @AkhilSrinivasSolidigm Hopefully this solves the issue.

**Comment by AkhilSrinivasSolidigm** [comment]
@starry12 Thank you for your comment. 1) fix resolved the issue we were facing. On a side note, We did face another issue while recompiling Nvidia Driver Kernel Symbols with 5.8 kernel with gcc version mismatch, although was ignored and built, caused cmake build errors. We did have to recompile the kernel with the current gcc version and this worked for us. Thank you again @starry12
